### DISCUSSION TITLE
Implicit parameter modifiers with delegates

### DISCUSSION THREAD

**Discussion by JamesYFC** [dotnet/csharplang#4913] [Language Ideas] [answered]
From [this](https://stackoverflow.com/questions/68311567/implicit-parameter-modifiers-with-delegates-c/68311973?noredirect=1#comment120743190_68311973) Stack Overflow question

If I define a delegate with some parameter modifiers, then I can't instantiate that delegate later on with implicit parameters. 

Example from SO:
```cs
public struct SomeStruct 
{ 
    public string x; 
    // and possibly more...
}

public delegate void MyDelegate(object foo, in SomeStruct bar);
```

However, when I go to create it, it tells me I need to put `in` there.

```cs
// Parameter 2 must be declared with the 'in' keyword
MyDelegate x = (foo, bar) => System.Console.WriteLine(bar.x);
```

If I use `in`, I now have to explicitly type the parameter...

```cs
// doesn't work
MyDelegate x = (foo, in bar) => System.Console.WriteLine(bar.x);
```

But now that I explicitly type the second parameter, the first parameter needs to be explicit too.

```cs
// Inconsistent lambda parameter usage; parameter types must be all explicit or all implicit
MyDelegate x = (foo, in SomeStruct bar) => System.Console.WriteLine(bar.x);
```

```cs
// ok
MyDelegate x = (object foo, in SomeStruct bar) => System.Console.WriteLine(bar.x);
```

The parameters `foo` and `bar` could have long type names, generic parameters, etc.

The type information is already there where this delegate type is defined, so I would like to avoid having to explicitly type the parameters for this delegate every time it's instantiated.

I was hoping I could write `(foo, bar) =>` or `(foo, in bar) =>`. 

Is this a feasible change?

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/issues/338.

**Reply by JamesYFC** [reply]
Cheers. Will be following that issue
